elierotenberg/learning-planet
# Add "Audit" and "Profile" pages, improve UI/UX

1) New features

* Client layout is now more or less responsive and more visually appealing
* Sidebar now provides quick access to pages in both logged-out and logged-in
  modes
* Add "Audit" screen to logged-in mode, which allows to audit past events, active
  tokens, and revoke specific tokens unilaterally
* Add "Profile" screen to logged-in mode, which allows to edit certain parts
  of the user profile, including creating/removing/updating its identity profile

2) Schema/API changes

* add AuditEvent concept in the database and in the OpenAPI definition
* add AuditToken concept in the OpenAPI definition and add denorm fields
  to the open_id_token table in the Database to ease querying/retrieving data

* add local API endpoints for interacting with the introduced features:
  - GET /local/me/userinfo
    Retrieve profile information about the current user
  - PUT /local/me/userinfo
    Update profile information about the current user
  - GET /local/me/audit
    Retrieve audit information about the current user
  - DELETE /local/me/token/{token_id}
    Revoke a token belonging to the current user

3) Development notes

* typescript-eslint in VSCode doesn't work well with eslint@^6 yet,
  so several temporary fixes have been applied:
    - add SKIP_PREFLIGHT_CHECK directive to client .env file
      because CRA expects latest (@^6) eslint but we can't use it;
      this should be removed whenever the incompability is fixed.
    - specify eslint working directories in VSCode workspace settings
      to mitigate some pathing issues

* during development, CSRF tokens and baseURL can be injected either:
  - by the server
  - in the URL query
  - as REACT_APP_ environment variables (at build time)
  - interactively in a native prompt

* client package:
  - better usage of SCSS modules
  - contextify parts of the global app state, including:
    * alerts
    * session
    * local API client including session state
  - componentize several common UI features:
    * page layout
    * vertical sidebar
    * horizontal button bar
    * alerts

* database package:
  - add more debug tooling in the Client class, such as query logging
  - fix critical bug in the expired tokens deletion query (was actually
    deleting the opposite - the non-expired tokens)

* interface package:
  - no changes

* server package:
  - change dev/local nonce byte length to 8 instead of 512; this is virtually the
  same security level but makes for much lighter queries (notably headers size)
  - add some QoL config options:
    * attempt to run a migration on startup
    * wait until a given service is reachable on the network
      This is especially useful when using Docker Compose or another orchestration tool;
      using this option, the server can wait until an upstream service is reachable
      and ready to accept connections before attempting to start
    * verbose logs in tests only (ignored unless NODE_ENV = test)
  - add more TestUtils, available only if NODE_ENV = test, including several Puppeter related
    utils to dedupe code
  - rename OpenIdProvider class to RouteHandler as it is actually a route handler which
    happens to implement the endpoints required to be an OpenID Provider
    More refactor could prove useful as the App and RouteHandler class are now pretty
    much god objects
